### PR TITLE
JavaScript: Improve detection of UMD modules.

### PR DESCRIPTION
**Pull request by max-schaefer**
Specifically, we now detect this pattern:

```js
(function (global, factory) {
    checkForNodeJS() ? module.exports = factory(...) :
    checkForAmd() ? define(..., factory) :
    global.mymodule = factory(...)
}(this, function (...) { ... });
```

We previously missed it because we expected the `define` to appear as an expression statement, but here it is nested a bit more deeply.

I fixed this by using the `inVoidContext` predicate we use in a few other queries, which however needed a bit of extra work. I'm curious to see whether that has any knock-on effects on performance or results, if it does I'm happy to switch to a more targeted fix.

### CODE REVIEW

File: javascript/ql/lib/Expressions/ExprHasNoEffect.qll [hunk 1]
```diff
@@ -38,6 +38,8 @@ predicate inVoidContext(Expr e) {
   )
   or
   exists(LogicalBinaryExpr logical | e = logical.getRightOperand() and inVoidContext(logical))
+  or
+  exists(ConditionalExpr cond | e = cond.getABranch())
```
Comment (erik-krogh) [inline javascript/ql/lib/Expressions/ExprHasNoEffect.qll:42]: Don't you need to add `and inVoidContext(cond)`?

Comment (max-schaefer) [reply javascript/ql/lib/Expressions/ExprHasNoEffect.qll:42]: 🙈

Comment (max-schaefer) [reply javascript/ql/lib/Expressions/ExprHasNoEffect.qll]: That would indeed have had a bit of an effect on results! Now fixed.

**Review by erik-krogh — APPROVED** [review]
LGTM assuming an evaluation works out.  
(Let me know if you want me to do one for you).

**Comment by max-schaefer** [comment]
> Let me know if you want me to do one for you

That would be fantastic, thanks! (The language tests pass.)

**Comment by erik-krogh** [comment]
> > Let me know if you want me to do one for you
> 
> That would be fantastic, thanks! (The language tests pass.)

I've started one. It should appear as a backref shortly.

**Comment by max-schaefer** [comment]
The evaluation revealed significant performance regressions that need looking into.

**Comment by max-schaefer** [comment]
I took a quick look at the regression, but didn't make much progress. It's easy to reproduce by, for example, running `ClientSideUrlRedirect.ql` on `openlayers/ol2`, which goes from 2s to 105s. We seem to be evaluating a whole load of predicates we previously didn't evaluate (or perhaps they were very quick, I don't know how to read logs anymore these days). I can't justify spending more time on this right now, so I'll just leave this open as a draft in case anyone else wants to pick it up later.

**Comment by erik-krogh** [comment]
I've looked a bit at this, and I think I we could merge it now.  

[New evaluation](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/umd-default-code-scanning__1/reports).  

Some of the performance regression seem to have disappeared from just waiting (core improvements?).    
But there is still a performance regression left.  

However, the cause seems to be that we recognize more flowsteps.   
There are several new results, and all the new results (I sampled a few) seem to be from us having a more complete callgraph, which is caused from recognizing more UMD modules.   
Some of those new results are eventually FPs (e.g. flow through `_.escape(..)` in `adobe/brackets`), but the new edge in the callgraph is good.  

I think the trade-off is worth it.

**Comment by max-schaefer** [comment]
> I think we could merge it now

Ah, self-fixing performance problems! Those are the best.

**Review by asgerf — APPROVED** [review]

**Pull request merged**